log/sequencer.go

```diff
@@ -239,141 +239,140 @@ func (s Sequencer) SequenceBatch(ctx context.Context, logID int64, limit int, gu
-	guardCutoffTime := s.timeSource.Now().Add(-guardWindow)
-	leaves, err := tx.DequeueLeaves(ctx, limit, guardCutoffTime)
-	if err != nil {
-		glog.Warningf("%v: Sequencer failed to dequeue leaves: %v", logID, err)
-		return 0, err
-	}
-	seqDequeueLatency.Observe(s.since(stageStart), label)
-	stageStart = s.timeSource.Now()
+	numLeaves := 0
+	var newLogRoot *trillian.SignedLogRoot
+	err := s.logStorage.ReadWriteTransaction(ctx, logID, func(ctx context.Context, tx storage.LogTreeTX) error {
+		defer seqBatches.Inc(label)
+		defer func() { seqLatency.Observe(s.since(start), label) }()
+
+		// Very recent leaves inside the guard window will not be available for sequencing
+		guardCutoffTime := s.timeSource.Now().Add(-guardWindow)
+		leaves, err := tx.DequeueLeaves(ctx, limit, guardCutoffTime)
+		if err != nil {
+			glog.Warningf("%v: Sequencer failed to dequeue leaves: %v", logID, err)
+			return err
+		}
+		seqDequeueLatency.Observe(s.since(stageStart), label)
+		stageStart = s.timeSource.Now()
 
-	// Get the latest known root from storage
-	currentRoot, err := tx.LatestSignedLogRoot(ctx)
-	if err != nil {
-		glog.Warningf("%v: Sequencer failed to get latest root: %v", logID, err)
-		return 0, err
-	}
-	seqGetRootLatency.Observe(s.since(stageStart), label)
-	stageStart = s.timeSource.Now()
-
-	// TODO(al): Have a better detection mechanism for there being no stored root.
-	// TODO(mhs): Might be better to create empty root in provisioning API when it exists
-	if currentRoot.RootHash == nil {
-		glog.Warningf("%v: Fresh log - no previous TreeHeads exist.", logID)
-		// SignRoot starts a new transaction, and we've got one open here until
-		// this function returns.
-		// This explicit Close() is a work-around for the in-memory storage which
-		// locks the tree for each TX.
-		// TODO(al): Producing the first signed root for a new tree should be
-		// handled by the provisioning, move it there.
-		tx.Close()
-		return 0, s.SignRoot(ctx, logID)
-	}
+		// Get the latest known root from storage
+		currentRoot, err := tx.LatestSignedLogRoot(ctx)
+		if err != nil {
+			glog.Warningf("%v: Sequencer failed to get latest root: %v", logID, err)
+			return err
+		}
+		seqGetRootLatency.Observe(s.since(stageStart), label)
+		stageStart = s.timeSource.Now()
+
+		// TODO(al): Have a better detection mechanism for there being no stored root.
+		// TODO(mhs): Might be better to create empty root in provisioning API when it exists
+		if currentRoot.RootHash == nil {
+			glog.Warningf("%v: Fresh log - no previous TreeHeads exist.", logID)
+			// SignRoot starts a new transaction, and we've got one open here until
+			// this function returns.
+			// This explicit Close() is a work-around for the in-memory storage which
+			// locks the tree for each TX.
+			// TODO(al): Producing the first signed root for a new tree should be
+			// handled by the provisioning, move it there.
+			tx.Close()
+			return s.SignRoot(ctx, logID)
```

Comment:
Is this block still safe?  Won't the wrapper mean that `Close()` gets called twice?

Maybe now is the time to do the TODO above and make this a provisioning function (and I think `SignRoot` below can then be dropped).

Reply:
I thought provisioning did create log roots? I heard this claim being made yesterday.

Reply:
This overlaps with the changes to split out sequencing. Merge might be tricky if you're unlucky.

Reply:
I'm pretty sure `Close` is idempotent.

I've just done explicit init for the Map, not opposed to do that for the Log too, but I'm trying to keep this PR focused on the TX change.

I don't see any code in `admin_server.CreateTree` which could make a new root...

Reply:
Even if `Close` is idempotent it still makes me very nervous to have a rarely-tested arm inside a 'run-in-transaction' method that closes off the enclosing transaction and runs a different transaction instead.

Reply:
This has been sorted out with #972